Lib/email/feedparser.py

```diff
@@ -87,40 +87,261 @@ def readline(self):
-        if '\n' not in data and '\r' not in data:
+        if not data:
+            return
+
+        if self._can_dump_data(data):
+            self._dump_destination.append(data)
+            return
+
+        self._push_data(data)
+
+    def _can_dump_data(self, data):
+        if self._dump_destination is None:
+            return False
+
+        # We're dumping; check for easy optimizations
+        if not self._eofstack:
+            # There's nothing that will ever tell us to stop dumping.
+            # This does absolute wonders for large non-multipart emails.
+            assert not self._lines
+            assert not self._dangling_partial
+            assert not self._partial
+            return True
+
+        # We can't dump this blob if we have pending partial data
+        if self._partial:
+            return False
+
+        for pred in self._eofstack:
+            if not hasattr(pred, 'is_boundary_match'):
+                # We can't blindly dump entire chunks, if we're interested in
+                # more than just boundaries
+                return False
+
+        # We only care about boundaries; we can dump as long as there's no
+        # potential boundaries.
+        return '-' not in data
+
+    def _can_dump_partial(self, line, start=0, end=sys.maxsize):
+        # Very similar to _can_dump_data above, except we can make some
+        # additional assumptions for partials/lines.
+        assert not self._partial or line is self._partial[0]
+
+        if self._dump_destination is None:
+            return False
+
+        # We're dumping. There should be absolutely no other pending lines,
+        # because those should've been dumped.
+        assert not self._lines
+        if not self._eofstack:
+            # There's nothing that will ever tell us to stop dumping. Dump away
+            return True
+
+        all_boundary_matches = True
+        for pred in self._eofstack:
+            if not hasattr(pred, 'is_boundary_match'):
+                return False
+
+        # We only care about boundaries; we can dump as long as there's no
+        # potential boundaries.
+        return not line.startswith("-", start, end)
+
+    def _is_dump_midline(self):
+        if not self._dump_destination:
+            return False
+
+        return self._dump_destination[-1][-1] not in ('\n', '\r')
+
+    def _push_data(self, data):
+        # Find first newline character in the data
+        unl_start_index = BufferedSubFile._find_unl(data)
+        if unl_start_index < 0:
             # No new complete lines, wait for more.
+            # Check to see if we had a previous dangling partial newline
+            if self._dangling_partial:
+                # We previously pushed a dangling line expecting \n to follow,
+                # however we received other data instead. Therefore, that \r
+                # does actually terminate a line. Go ahead and push it.
+                self._flush_partial()
```

Comment:
One way to reduce indentation, and make it easier to read perhaps, is to actually handle `self._dangling_partial` in `flush_partial` directly, or in another method, say `self._try_flush_partial()`.